rd-net/RdFramework/Impl/Serializers.cs

```diff
@@ -348,24 +324,75 @@ bool TryGetTypeMapping(Type type1, out RdId rdId)
     }
 
     
-    private readonly HashSet<Type> myRegisteredToplevels = new HashSet<Type>();
     public void RegisterToplevelOnce(Type toplevelType, Action<ISerializers> registerDeclaredTypesSerializers)
     {
-      var r = new ToplevelRegistration(toplevelType, registerDeclaredTypesSerializers);
-#if !NET35
-      var task = myBackgroundRegistrar.SendOrExecuteInline(r);
-      Assertion.Assert(task.IsCompleted, "task.IsCompleted: {0}", task.Status);
-#else
-      RegisterToplevelInternal(r);
- #endif
+      myBackgroundRegistrar.Queue(toplevelType, registerDeclaredTypesSerializers);
+    }
+
+    struct ToplevelRegistration
+    {
+      internal Type Type { get; }
+      internal Action<ISerializers> Action { get; }
+
+      public ToplevelRegistration(Type type, Action<ISerializers> action)
+      {
+        Type = type;
+        Action = action;
+      }
     }
 
-    private void RegisterToplevelInternal(ToplevelRegistration r)
+    private class BackgroundRegistrar
     {
-      if (!myRegisteredToplevels.Add(r.Type)) return;
-      Protocol.InitTrace?.Log($"REGISTER serializers for {r.Type.Name}");
+      [NotNull] private readonly Serializers myOwner;
+      [NotNull] private readonly TaskScheduler myScheduler;
+
+      [NotNull] private readonly object myLock = new object();
+      [NotNull] private readonly Queue<ToplevelRegistration> myQueue = new Queue<ToplevelRegistration>();
+      [NotNull] private readonly HashSet<Type> myRegisteredToplevels = new HashSet<Type>();
+
+      private int myQueueSize;
+
+      public BackgroundRegistrar([NotNull] Serializers owner, [CanBeNull] TaskScheduler scheduler)
+      {
+        myOwner = owner;
+        myScheduler = scheduler ?? TaskScheduler.Default;
+      }
+
+      public void Queue(Type type, Action<ISerializers> action)
+      {
+        lock (myLock)
+        {
+          if (!myRegisteredToplevels.Add(type))
+            return;
 
-      r.Action(this);
+          myQueueSize++;
+          myQueue.Enqueue(new ToplevelRegistration(type, action));
+        }
+
+        var task = new Task(ProcessQueue);
+        task.Start(myScheduler);
+      }
+
+      public void ProcessQueue()
+      {
+        if (myQueueSize == 0)
```

Comment:
Is it safe to read a non-volatile variable from outside of a lock? I'd recommend to declare `myQueueSize` as `volatile`.

Reply:
all primitive types up to 32 bits is safe to read outside of a lock (and get stale value sometimes, but it is not an issue here)

Reply:
Only on x86, IIRC?

Reply:
I'm not aware of such restrictions and could not google it. I don't mind adding it, by the way (and I've added it).
I just wanted to understand in detail what exactly we are protecting from, and not adhere to some cargo-cult.